lib/runtime/errors/transaction_validity.go

```diff
@@ -0,0 +1,105 @@
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// TransactionValidityError Information on a transaction's validity and, if valid,
+// on how it relates to other transactions. It is a result of the form:
+// Result<ValidTransaction, TransactionValidityError>
+type TransactionValidityError scale.VaryingDataType
+
+var (
+	errInvalidType     = errors.New("invalid validity type")
+	errInvalidResult   = errors.New("invalid error value")
+	errInvalidTypeCast = errors.New("invalid type cast")
+)
+
+// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
+func (tve *TransactionValidityError) Set(val scale.VaryingDataTypeValue) (err error) {
+	vdt := scale.VaryingDataType(*tve)
+	err = vdt.Set(val)
+	if err != nil {
+		return err
+	}
+	*tve = TransactionValidityError(vdt)
+	return nil
+}
+
+// Value will return the value from the underlying VaryingDataType
+func (tve *TransactionValidityError) Value() (val scale.VaryingDataTypeValue) {
+	vdt := scale.VaryingDataType(*tve)
+	return vdt.Value()
+}
+
+// Error will return the error underlying TransactionValidityError
+func (tve *TransactionValidityError) Error() string {
+	invalidTxn, ok := tve.Value().(InvalidTransaction)
+	if !ok {
+		unknownTxn, ok2 := tve.Value().(UnknownTransaction)
+		if !ok2 {
+			return errInvalidTypeCast.Error()
+		}
+		return unknownTxn.Error()
+	}
+	return invalidTxn.Error()
+}
+
+// NewTransactionValidityError is constructor for TransactionValidityError
+func NewTransactionValidityError() TransactionValidityError {
+	vdt, err := scale.NewVaryingDataType(NewInvalidTransaction(), NewUnknownTransaction())
+	if err != nil {
+		panic(err)
+	}
+	return TransactionValidityError(vdt)
+}
+
+var (
+	ErrInvalidTxn = errors.New("invalid transaction")
+	ErrUnknownTxn = errors.New("unknown transaction")
+)
+
+// UnmarshalTransactionValidity takes the result of the validateTransaction runtime call and unmarshalls it
+// TODO use custom result issue #2780
+func UnmarshalTransactionValidity(res []byte) (*transaction.Validity, error) {
+	validTxn := transaction.Validity{}
+	txnValidityErrResult := NewTransactionValidityError()
+	txnValidityResult := scale.NewResult(validTxn, txnValidityErrResult)
+	err := scale.Unmarshal(res, &txnValidityResult)
+	if err != nil {
+		return nil, fmt.Errorf("scale decoding transaction validity result: %w", err)
+	}
+	txnValidityRes, err := txnValidityResult.Unwrap()
+	if err != nil {
+		scaleWrappedErr, ok := err.(scale.WrappedErr)
+		if ok {
+			txnValidityErr, ok := scaleWrappedErr.Err.(TransactionValidityError)
+			if !ok {
+				return nil, fmt.Errorf("%w: %T", errInvalidTypeCast, scaleWrappedErr.Err)
+			}
+
+			switch txnValidityErr.Value().(type) {
+			// TODO use custom result issue #2780
+			case InvalidTransaction:
+				return nil, fmt.Errorf("%w: %s", ErrInvalidTxn, txnValidityErr.Error())
+			case UnknownTransaction: // do nothing
+				return nil, fmt.Errorf("%w: %s", ErrUnknownTxn, txnValidityErr.Error())
```

Comment:
you are returning the same error in this case?

Reply:
The sentinel error is different, but the stringified error string is the same var (well it's different obviously). We just stringify it to add context to the error returned.

Reply:
IMO we don't need a sentinel error in this case. We have types that fulfil the `error` interface so why not use them.  They are distinguishable since they are concrete types.